draft-ietf-quic-load-balancers.md

```diff
@@ -752,20 +767,108 @@ such as congestion window, RTT, or MTU. Opaque data MAY also allow servers to
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| AEAD Encrypted Token (variable)                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| AEAD Checksum (variable, function of encryption algorithm)    |                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+
+The tokens are protected using AES128-GCM as follow:
+
+* The token key and IV are retrieved using the Key Sequence.
+
+* The nonce, N, is formed by combining the IV with the 32 bit server ID 
+and the 64 bit token
+sequence number, producing a 96 bit unique token number.
+The 96 bits of the unique token number are left-padded
+with zeros to the size of the IV. The exclusive OR of the padded token sequence
+number and the IV forms the AEAD nonce.
+
+* The associated data is a formatted as a pseudo header by 
+  combining the clear-text part of the token with the IP address
+  of the client.
+~~~~~
+0                   1                   2                   3
+0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                                                               |
++                                                               +
+|                                                               |
++                        IP Address (128)                       +
+|                                                               |
++                                                               +
+|                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                     Server ID (32)                            |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                                                               |
++                 Token sequence number (64)                    +
+|                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| Key seq.(8)   |
++-+-+-+-+-+-+-+-+
+~~~~~
+* The input plaintext for the AEAD is the token body. The output ciphertext
+of the AEAD is transmitted in place of token body.
+* The AEAD Checksum is computed as part of the AEAD encryption process,
+and is verified during decryption.
+
 ### Configuration Agent Actions
 
-The configuration agent generates and distributes a "token key" and a list of
+The configuration agent generates and distributes a "token key", a "token IV",
+a key sequence,
+and a list of
 QUIC versions the service supports. It must also inform the service if a NAT
 lies between the service and the servers.
 
+The configuration also specifies whether the unique token number is
+generated using the "random" or the "controlled" methods. If the controlled
+method is used, the configuration agent provisions each server with a
+unique 32 bit server ID.
```

Comment:
The Retry Service also needs a "generator ID". Arguably, it would make sense to have more than one (it presumably will generate tokens in much higher volume).

Reply:
Disregard, no server IDs anymore